File: backend/Evervault.Api/Services/Ai/AiTypes.cs
namespace Evervault.Api.Services.Ai;

/// <summary>Neutral chat message exchanged with any provider. The client holds the transcript and
/// re-sends it each turn (stateless backend).</summary>
public record AiChatMessage(
    string Role,                          // "system" | "user" | "assistant" | "tool"
    string? Content = null,
    List<AiToolCall>? ToolCalls = null,   // present on assistant turns that called tools
    string? ToolCallId = null,            // present on tool-result turns
    string? Name = null,                  // tool name (on tool-result turns)
    string? ProviderState = null);        // opaque provider-only state round-tripped via the client
                                          // transcript (OpenAI: reasoning items to re-send under store:false)

/// <summary>A tool/function call the model wants to make. Arguments stay as a raw JSON string.
/// <see cref="ThoughtSignature"/> is Gemini 3.x's part-level signature — it must be echoed verbatim
/// when the call is replayed on the next turn or the request 400s (other providers leave it null).</summary>
public record AiToolCall(string Id, string Name, string ArgumentsJson, string? ThoughtSignature = null);

/// <summary>Token usage for one model round-trip, where the provider reports it (all nullable — some
/// providers/paths don't expose every field). Used only for logging/observability.</summary>
public record AiUsage(int? PromptTokens, int? CompletionTokens, int? TotalTokens);

/// <summary>One model round-trip result: either assistant text, tool calls, or both. <see cref="ProviderState"/>
/// is optional opaque data the provider needs echoed on the next turn (see <see cref="AiChatMessage"/>).
/// <see cref="Usage"/> carries token counts when the provider reported them (best-effort, for logging).</summary>
public record AiCompletion(string? Text, List<AiToolCall> ToolCalls, string? ProviderState = null, AiUsage? Usage = null);

/// <summary>A provider-agnostic model entry for the switcher. <see cref="ReasoningLevels"/> (when the
/// provider advertises them, e.g. ChatGPT) lets the UI offer only the effort levels this model supports.</summary>
public record AiModelInfo(
    string Id,
    string Name,
    string Provider,
    bool IsFree,
    decimal? PromptPricePerMTok,
    decimal? CompletionPricePerMTok,
    string? PriceLabel,
    IReadOnlyList<string>? ReasoningLevels = null,
    string? DefaultReasoningLevel = null);

/// <summary>A tool the model may call, in provider-agnostic form. <see cref="ParametersJson"/> is a
/// JSON-Schema string describing the arguments.</summary>
public record AiToolSchema(string Name, string Description, string ParametersJson);

/// <summary>One source behind a grounded search answer. <see cref="Uri"/> is exactly what the provider
/// returned — for Gemini that is an opaque <c>grounding-api-redirect</c> link, NOT the real page — so it
/// must be resolved before it can be shown to anyone (see <c>GeminiWebSearchService</c>).
/// <see cref="Title"/> is typically the bare domain rather than the page's headline.</summary>
public record GroundedSearchSource(string Title, string Uri, string Snippet);

/// <summary>The result of one search-grounded generation: the model's synthesized answer plus the web
/// sources it drew on. <see cref="Sources"/> is empty when the model chose to answer without searching,
/// which is a normal outcome rather than an error.</summary>
public record GroundedSearch(string? Answer, IReadOnlyList<GroundedSearchSource> Sources, AiUsage? Usage);

/// <summary>Per-call generation tuning, provider-agnostic. Today only carries reasoning effort;
/// add temperature/max-tokens/etc. here later. <see cref="ReasoningEffort"/> is one of
/// "auto" | "off" | "low" | "medium" | "high" (null/"auto" → providers send nothing special).</summary>
public record AiGenerationOptions(string? ReasoningEffort = null);

/// <summary>Credit/quota usage for a key, where the provider exposes it (OpenRouter does; Gemini does not).
/// The Daily*/ResetUnixMs fields come from rate-limit response headers (best-effort) and describe the
/// free-model daily request allowance.</summary>
public record AiKeyUsage(
    bool Supported,
    string? Summary,
    decimal? Usage,
    decimal? Limit,
    decimal? Remaining,
    bool? IsFreeTier,
    string? RateLimit,
    string? ResetNote,
    long? DailyLimit = null,
    long? DailyRemaining = null,
    long? DailyUsed = null,
    long? ResetUnixMs = null,
    IReadOnlyList<AiRateWindow>? Windows = null);

/// <summary>One rolling rate-limit window (e.g. ChatGPT's 5-hour and weekly quotas): how much of the
/// window is consumed and when it resets. <see cref="ResetUnixMs"/> is epoch milliseconds.</summary>
public record AiRateWindow(string Label, double UsedPercent, long? ResetUnixMs);

public enum AiErrorKind { Auth, Quota, Transient, Other }

/// <summary>A typed provider error. <see cref="Kind"/> drives key failover (Auth/Quota/Transient
/// advance to the next key; Other is surfaced immediately).</summary>
public class AiProviderException : Exception
{
    public AiErrorKind Kind { get; }
    public AiProviderException(AiErrorKind kind, string message) : base(message) => Kind = kind;
}

/// <summary>Thrown when every key for a provider failed. Carries one message per key.</summary>
public class AllKeysFailedException : Exception
{
    public IReadOnlyList<string> Errors { get; }
    public AllKeysFailedException(IReadOnlyList<string> errors)
        : base("All API keys failed.") => Errors = errors;
}

/// <summary>A chat/model provider (OpenRouter, Gemini, …). Implementations are stateless and take the
/// raw key per call — the <see cref="KeyFailoverRunner"/> owns key selection.</summary>
public interface IAiProvider
{
    string Name { get; }
    Task<(bool Ok, string Message)> ValidateKeyAsync(string rawKey, CancellationToken ct);
    Task<IReadOnlyList<AiModelInfo>> ListModelsAsync(string rawKey, CancellationToken ct);
    /// <summary>List models for a purpose: "chat" (generateContent) or "embedding" (embedContent).
    /// Defaults to the chat list for providers that don't distinguish.</summary>
    Task<IReadOnlyList<AiModelInfo>> ListModelsAsync(string rawKey, string kind, CancellationToken ct)
        => ListModelsAsync(rawKey, ct);
    Task<AiKeyUsage> GetUsageAsync(string rawKey, CancellationToken ct);
    Task<AiCompletion> CompleteAsync(
        string rawKey,
        string model,
        IReadOnlyList<AiChatMessage> messages,
        IReadOnlyList<AiToolSchema> tools,
        AiGenerationOptions? options,
        CancellationToken ct);

    /// <summary>Like <see cref="CompleteAsync"/> but reports assistant text incrementally through
    /// <paramref name="onTextDelta"/> while the provider streams; the returned completion is still the
    /// full aggregate (text, tool calls, provider state, usage). Default: no incremental support — the
    /// callback is never invoked and the caller must fall back to <see cref="AiCompletion.Text"/>.</summary>
    Task<AiCompletion> CompleteStreamingAsync(
        string rawKey,
        string model,
        IReadOnlyList<AiChatMessage> messages,
        IReadOnlyList<AiToolSchema> tools,
        AiGenerationOptions? options,
        Func<string, Task>? onTextDelta,
        CancellationToken ct)
        => CompleteAsync(rawKey, model, messages, tools, options, ct);

    /// <summary>Synthesize speech for the given text+voice. Returns raw PCM bytes and the audio mime
    /// (e.g. "audio/L16;codec=pcm;rate=24000"). Default: the provider has no TTS — throws Other so
    /// <see cref="KeyFailoverRunner"/> surfaces it immediately instead of retrying every key.</summary>
    Task<(byte[] Pcm, string Mime)> SynthesizeSpeechAsync(
        string rawKey, string model, string text, string voiceName, CancellationToken ct)
        => throw new AiProviderException(AiErrorKind.Other, "This provider does not support TTS.");
}

public interface IAiProviderFactory
{
    IAiProvider Get(string provider);

    /// <summary>True when AI_FAKE=1 — every provider resolves to the offline <see cref="FakeAiProvider"/>.
    /// Lets other seams (e.g. <see cref="KeyFailoverRunner"/>) skip real credential lookups.</summary>
    bool IsFake { get; }
}

File: web/src/app/admin/aiTypes.ts
// Shared types for the admin AI features. Field names match the backend DTOs (camelCase).

export type Provider = "gemini" | "openrouter" | "openai";

export type AiKeyDto = {
  id: number;
  provider: string;
  keyHint: string;
  sortOrder: number;
  enabled: boolean;
};

export type AiKeysDto = {
  gemini: AiKeyDto[];
  openRouter: AiKeyDto[];
};

// Validity is checked on demand and shown only in the page — never persisted.
export type KeyCheckResult = { id: number; keyHint: string; ok: boolean; message: string };
export type CheckKeysResult = { results: KeyCheckResult[] };

export type ModelInfo = {
  id: string;
  name: string;
  provider: string;
  isFree: boolean;
  promptPricePerMTok: number | null;
  completionPricePerMTok: number | null;
  priceLabel: string | null;
  // Reasoning-effort levels this model supports (ChatGPT advertises these; others are null).
  reasoningLevels: string[] | null;
  defaultReasoningLevel: string | null;
};

export type ModelsResult = { provider: string; models: ModelInfo[]; warning: string | null };

// Models the admin picks for the public /webapp (keyless): text, TTS (voice messages), and the
// realtime live call, plus the default voice. Mirrors the backend WebappAiConfigDto.
// The text model has a primary + optional fallback, each Gemini or ChatGPT (with a reasoning level for
// ChatGPT). textModel is the primary model id; textProvider says which provider it belongs to.
export type WebappAiConfigDto = {
  textModel: string;
  audioModel: string;
  liveModel: string;
  defaultVoice: string;
  textProvider: string; // "gemini" | "openai" (primary). Defaults to "gemini".
  textReasoning: string | null; // primary reasoning level (ChatGPT only); null = model default
  textFallbackProvider: string | null; // "gemini" | "openai" | null (no fallback)
  textFallbackModel: string | null;
  textFallbackReasoning: string | null; // fallback reasoning level (ChatGPT only); null = model default
  // How long a live voice call may sit in user silence before it auto-hangs-up, in seconds. 0 = never.
  liveIdleTimeoutSeconds: number;
  // Voice-message replies: which Gemini Live model powers the "live" path, and the mode ("live" = one
  // Gemini Live session with automatic TTS fallback; "tts" = the legacy synthesis pipeline).
  voiceLiveModel: string;
  voiceMode: string; // "live" | "tts"
  // Thinking level for each Gemini Live leg: "minimal" | "low" | "medium" | "high", or null for the
  // model's default (Live defaults to minimal, for latency). Sent to the browser, which applies it to
  // the Live socket's thinkingConfig. The two are independent — a call can stay fast while a voice
  // message thinks harder.
  liveReasoning: string | null;
  voiceLiveReasoning: string | null;
};

// One rolling quota window (e.g. ChatGPT's 5-hour and weekly limits). resetUnixMs is epoch ms.
export type AiRateWindow = { label: string; usedPercent: number; resetUnixMs: number | null };

export type AiKeyUsage = {
  supported: boolean;
  summary: string | null;
  usage: number | null;
  limit: number | null;
  remaining: number | null;
  isFreeTier: boolean | null;
  rateLimit: string | null;
  resetNote: string | null;
  dailyLimit: number | null;
  dailyRemaining: number | null;
  dailyUsed: number | null;
  resetUnixMs: number | null;
  windows: AiRateWindow[] | null;
};

export type ChatConfigDto = {
  selectedProvider: string | null;
  geminiModel: string | null;
  openRouterModel: string | null;
  openAiModel: string | null;
  reasoningEffort: string | null; // "auto" | "off" | "low" | "medium" | "high" | null (Gemini/OpenRouter)
  openAiReasoning: string | null; // ChatGPT's own level (minimal/low/medium/high/xhigh/…), or "auto"
};

// "Sign in with ChatGPT" (Codex OAuth) connection status. No secrets.
export type OpenAiStatus = {
  connected: boolean;
  email: string | null;
  connectedAt: string | null;
  expiresAt: string | null;
};

export type EmbeddingConfigDto = {
  provider: string;
  model: string | null;
  dimensions: number;
  locked: boolean;
  lockedAt: string | null;
};

export type ToolCall = { id: string; name: string; argumentsJson: string };

export type ChatMessage = {
  role: string; // "system" | "user" | "assistant" | "tool"
  content?: string | null;
  toolCalls?: ToolCall[] | null;
  toolCallId?: string | null;
  name?: string | null;
  // Opaque provider-only state (OpenAI reasoning items) round-tripped verbatim through the transcript.
  providerState?: string | null;
};

export type ProposedAction = {
  toolCallId: string;
  toolName: string;
  argumentsJson: string;
  humanSummary: string;
  dangerous: boolean;
  signature: string;
};

export type ChatTurnResponse = {
  status: "message" | "proposal" | "error";
  assistantText?: string | null;
  proposal?: ProposedAction | null;
  error?: string | null;
  messages: ChatMessage[];
};
